spec.html

```diff
@@ -6419,135 +6419,31 @@ <h1>GetGlobalObject ( )</h1>
-    <emu-note>
-      <p>Typically an ECMAScript implementation will have its Job Queues pre-initialized with at least one PendingJob and one of those Jobs will be the first to be executed. An implementation might choose to free all resources and terminate if the current Job completes and all Job Queues are empty. Alternatively, it might choose to wait for a some implementation specific agent or mechanism to enqueue new PendingJob requests.</p>
-    </emu-note>
-    <p>The following abstract operations are used to create and manage Jobs and Job Queues:</p>
+    <h1>Jobs and Host Operations to Enqueue Jobs</h1>
+    <p>A <dfn>Job</dfn> is an abstract operation that initiates an ECMAScript computation when no other ECMAScript computation is currently in progress. A Job abstract operation may be defined to accept an arbitrary set of job parameters.</p>
 
-    <emu-clause id="sec-enqueuejob" aoid="EnqueueJob">
-      <h1>EnqueueJob ( _queueName_, _job_, _arguments_ )</h1>
-      <p>The EnqueueJob abstract operation requires three arguments: _queueName_, _job_, and _arguments_. It performs the following steps:</p>
-      <emu-alg>
-        1. Assert: Type(_queueName_) is String and its value is the name of a Job Queue recognized by this implementation.
-        1. Assert: _job_ is the name of a Job.
-        1. Assert: _arguments_ is a List that has the same number of elements as the number of parameters required by _job_.
-        1. Let _callerContext_ be the running execution context.
-        1. Let _callerRealm_ be _callerContext_'s Realm.
-        1. Let _callerScriptOrModule_ be _callerContext_'s ScriptOrModule.
-        1. Let _pending_ be PendingJob { [[Job]]: _job_, [[Arguments]]: _arguments_, [[Realm]]: _callerRealm_, [[ScriptOrModule]]: _callerScriptOrModule_, [[HostDefined]]: *undefined* }.
-        1. Perform any implementation or host environment defined processing of _pending_. This may include modifying the [[HostDefined]] field or any other field of _pending_.
-        1. Add _pending_ at the back of the Job Queue named by _queueName_.
-        1. Return NormalCompletion(~empty~).
-      </emu-alg>
+    <p>Jobs are scheduled for execution by ECMAScript host environments with abstract operations such as HostEnqueuePromiseJob. Such operations accept two parameters: _job_, an abstract operation, and _arguments_, a list of ECMAScript values. These abstract operations schedule _job_ to be performed with _arguments_ at some future time. Their implementations must conform to the following requirements:</p>
+
+    <ul>
+      <li>At some future point in time, when there is no running execution context and the execution context stack is empty, the implementation must:
```

Comment:
Since the preamble to this `<ul>` ends with "Their implementations must conform to the following requirements:", it's tempting to think that "the implementation" here refers to the same thing, i.e. the (host's) implementation of a job-enqueing operation. But it could also mean what it usually means elsewhere in the spec (roughly, the ECMAScript engine).

But neither option entirely makes sense to me.
- The implementation of a job-enqueuing operation is basically just "append this job to a queue". What happens when the jobs *runs* isn't the responsibility of the enqueuing operation.
- If it's the ECMAScript engine doing the following steps (which, by itself, seems to make sense), how does the job get from a host-maintained queue to the engine? It seems like there needs to be a `RunJob` operation that the host can call to get the engine to do this.

Reply:
Also, it's unclear how the the execution context stack would ever be empty. Specifically, InitializeHostDefinedRealm pushes an execution context onto the stack; what would ever pop that context from the stack?

Reply:
> The implementation of a job-enqueuing operation is basically just "append this job to a queue". What happens when the jobs runs isn't the responsibility of the enqueuing operation.

This isn't correct. See e.g. https://html.spec.whatwg.org/multipage/webappapis.html#enqueuejob(queuename,-job,-arguments) for what can be involved here. It is not simple at all.

If you haven't familiarized yourself with @littledan's presentation on this PR, that might be a good way of getting some shared context.

> Also, it's unclear how the the execution context stack would ever be empty. Specifically, InitializeHostDefinedRealm pushes an execution context onto the stack; what would ever pop that context from the stack?

In HTML the stack is popped immediately upon creation. The stack-pushing is just a kind of weird way of getting multiple return values out of InitializeHostDefinedRealm(). See https://html.spec.whatwg.org/multipage/webappapis.html#creating-a-new-javascript-realm.

Reply:
> > The implementation of a job-enqueuing operation is basically just "append this job to a queue". What happens when the jobs runs isn't the responsibility of the enqueuing operation.
> 
> This isn't correct. See e.g. https://html.spec.whatwg.org/multipage/webappapis.html#enqueuejob(queuename,-job,-arguments) for what can be involved here. It is not simple at all, and the enqueuing operation is intimately tied to the event loop, which in turn works to give these guarantees.

Sorry, yes, the host can make this as complicated as it likes, I was looking at it from the implementation's/engine's point of view. (I think.) 

But even so, I'm still not sure which meaning of "the implementation" is intended.

> If you haven't familiarized yourself with @littledan's presentation on this PR, that might be a good way of getting some shared context.

Sure. [This one?](https://docs.google.com/presentation/d/1LLIIwZyQgm6rbAyWbRjFFSNKlA2DPqslpN8sVKjchjE/edit#slide=id.p)

> > Also, it's unclear how the the execution context stack would ever be empty. Specifically, InitializeHostDefinedRealm pushes an execution context onto the stack; what would ever pop that context from the stack?
> 
> In HTML the stack is popped immediately upon creation.

Sneaky! 

> The stack-pushing is just a kind of weird way of getting a return value out of InitializeHostDefinedRealm(). See https://html.spec.whatwg.org/multipage/webappapis.html#creating-a-new-javascript-realm.

Ah, so that fits with my suggestion that InitializeHostDefinedRealm() needs to return the Realm Record that it creates.

Reply:
I've read @littledan's presentation a few times now (assuming I got the right one). I don't think it answers any of the questions I've asked so far, but maybe answers some I haven't asked yet.

Thanks for the HTML links. One thing that hadn't quite clicked for me before was that an HTML user agent can manipulate the ES execution stack. (This feels a bit odd to me, since the ES spec doesn't mention that possibility when it describes the stack.)

Reply:
I wouldn't mind looking into cleaning up this aspect of layering further in a follow-on PR.

Reply:
"Clean up" as in refactor the spec so that the host doesn't have to directly manipulate the stack? That'd be nice. (In a follow-on PR, agreed.)